docs: Developer Guide (#1437)
tendermint/starport: docs/guide/scavenge/09-play.md

---
order: 9
---

> # Play

Reviewer: ```suggestion
# Play with your Blockchain
```

## Starting the Blockchain

To start your blockchain in development, run the following command:

```
starport chain serve
```

`serve` will build the chain's binary, initialize a data directory and start a node in development. Keep it running in one terminal window and open another one to execute commands.

## Creating a Scavenge

> Let's follow the instructions and submit a new scavenge. The first parameter we need is the `solution`, but probably we should also know what the actual quesiton is that our solution solves (our `description`). How about our challenge question be something family friendly like: `What's brown and sticky?`. Of course the only solution to this question is: `A stick`.

Reviewer: ```suggestion
Let's follow the instructions and submit a new scavenge. The first parameter you need is the `solution`, but probably you should also know what the actual question is that your solution solves (the `description`). How about the challenge question be something family friendly like: `What's brown and sticky?`. Of course the only solution to this question is: `A stick`.
```

> Next we should specify our `reward`. Let's give away `100token` as a reward for solving our scavenge.

Reviewer: ```suggestion
Next your should specify the `reward`. Let's give away `100token` as a reward for solving the scavenge.
```

> Now we have all the pieces needed to create our message. Let's piece them all together, adding the flag `--from` so the CLI knows who is sending it:

Reviewer: ```suggestion
Now you have all the pieces needed to create our message. Piece them all together, adding the flag `--from` so the CLI knows who is sending it:
```

    },
    "auth_info": {
      "signer_infos": [
        {
          "public_key": {
            "@type": "/cosmos.crypto.secp256k1.PubKey",
            "key": "ApRuim5kLByq9AqJJ9dEF5rFCkAbhIehEcPzSouM92p6"
          },
          "mode_info": { "single": { "mode": "SIGN_MODE_DIRECT" } },
          "sequence": "1"
        }
      ],
      "fee": { "amount": [], "gas_limit": "200000", "payer": "", "granter": "" }
    },
    "signatures": [
      "8W5MkgV8oWpB6UWRGVKuimfPyb1OutG8KPXTIneM6WIvy4YHToG3GUXFpUrh+CxPXmlDh5gIfeR4+nFfUuQXng=="
    ]
  },
  "timestamp": "2021-07-09T10:24:52Z"
}
```

## Querying For a List of Scavenges

```
scavenged q scavenge list-scavenge --output json
```

```json
{
  "Scavenge": [
    {
      "creator": "cosmos1wzgkalxjhaqtznrzzp0xy5jgkxx82xaa660jxh",
      "index": "2f9457a6e8fb202f9e10389a143a383106268c460743dd59d723c0f82d9ba906",
      "solutionHash": "2f9457a6e8fb202f9e10389a143a383106268c460743dd59d723c0f82d9ba906",
      "solution": "",
      "description": "What's brown and sticky?",
      "reward": "100token",
      "scavenger": ""
    }
  ],
  "pagination": { "next_key": null, "total": "0" }
}
```

## Committing a solution

> Let's follow the instructions and submit the answer as a commit on behalf of `bob`:

Reviewer: ```suggestion
Follow the instructions and submit the answer as a commit on behalf of `bob`:
```

```

```json
{
  "body": {
    "messages": [
      {
        "@type": "/cosmonaut.scavenge.scavenge.MsgCommitSolution",
        "creator": "cosmos1gkheudhhjsvq0s8fxt7p6pwe0k3k30kepcnz9p",
        "solutionHash": "2f9457a6e8fb202f9e10389a143a383106268c460743dd59d723c0f82d9ba906",
        "solutionScavengerHash": "461d54ec0bbb1d696a79af80d7f63e4c6df262d76309423da37189453eaec127"
      }
    ],
    "memo": "",
    "timeout_height": "0",
    "extension_options": [],
    "non_critical_extension_options": []
  },
  "auth_info": {
    "signer_infos": [],
    "fee": { "amount": [], "gas_limit": "200000", "payer": "", "granter": "" }
  },
  "signatures": []
}
```

## Querying For a List of Commits

```
scavenged q scavenge list-commit --output json
```

```json
{
  "Commit": [
    {
      "creator": "cosmos1gkheudhhjsvq0s8fxt7p6pwe0k3k30kepcnz9p",
      "index": "461d54ec0bbb1d696a79af80d7f63e4c6df262d76309423da37189453eaec127",
      "solutionHash": "2f9457a6e8fb202f9e10389a143a383106268c460743dd59d723c0f82d9ba906",
      "solutionScavengerHash": "461d54ec0bbb1d696a79af80d7f63e4c6df262d76309423da37189453eaec127"
    }
  ],
  "pagination": { "next_key": null, "total": "0" }
}
```

> We don't need to put the `solutionHash` because it can be generated by hashing our actual solution.

Reviewer: ```suggestion
You don't need to put the `solutionHash` because it can be generated by hashing the actual solution.
```

> Since all we need is the solution again let's send and confirm our final message:

Reviewer: ```suggestion
Since all you need is the solution again, send and confirm the final message:
```

## Revealing a Solution

```
scavenged tx scavenge reveal-solution "A stick" --from bob
```

```json
{
  "body": {
    "messages": [
      {
        "@type": "/cosmonaut.scavenge.scavenge.MsgRevealSolution",
        "creator": "cosmos1gkheudhhjsvq0s8fxt7p6pwe0k3k30kepcnz9p",
        "solution": "A stick"
      }
    ],
    "memo": "",
    "timeout_height": "0",
    "extension_options": [],
    "non_critical_extension_options": []
  },
  "auth_info": {
    "signer_infos": [],
    "fee": { "amount": [], "gas_limit": "200000", "payer": "", "granter": "" }
  },
  "signatures": []
}
```

> ## Querying For a List of Scavenges (solved)

Reviewer: ```suggestion
## Querying For a List of solved Scavenges
```

```
scavenged q scavenge list-scavenge --output json
```

```json
{
  "Scavenge": [
    {
      "creator": "cosmos1wzgkalxjhaqtznrzzp0xy5jgkxx82xaa660jxh",
      "index": "2f9457a6e8fb202f9e10389a143a383106268c460743dd59d723c0f82d9ba906",
      "solutionHash": "2f9457a6e8fb202f9e10389a143a383106268c460743dd59d723c0f82d9ba906",
      "solution": "A stick",
      "description": "What's brown and sticky?",
      "reward": "100token",
      "scavenger": "cosmos1gkheudhhjsvq0s8fxt7p6pwe0k3k30kepcnz9p"
    }
  ],
  "pagination": { "next_key": null, "total": "0" }
}
```

> ## Committing a Solution Again to Get an Error

Reviewer: ```suggestion
## Committing a Solution Again, expect to get an Error
```
